toolchains/simple-pnpm/macros.bzl

```diff
@@ -0,0 +1,103 @@
+
+load(":toolchain.bzl", "SimplePnpmToolchainInfo",)
+
+def build_node_modules_impl(ctx: AnalysisContext) -> list[DefaultInfo]:
+    out = ctx.actions.declare_output("root", dir = True)
+
+    simple_pnpm_toolchain = ctx.attrs._simple_pnpm_toolchain[SimplePnpmToolchainInfo]
+    package_dir = cmd_args(ctx.label.package).relative_to(ctx.label.cell_root)
+
+    cmd = cmd_args(
+        ctx.attrs._python_toolchain[PythonToolchainInfo].interpreter,
+        simple_pnpm_toolchain.build_node_modules[DefaultInfo].default_outputs,
+    )
+    cmd.add("--root-dir")
+    cmd.add(package_dir)
+
+    cmd.add(out.as_output())
+    cmd.hidden([ctx.attrs.pnpm_lock])
+
+    ctx.actions.run(cmd, category = "pnpm", identifier = "install")
+
+    return [DefaultInfo(default_output = out)]
+
+build_node_modules = rule(
+    impl = build_node_modules_impl,
+    attrs = {
+        "pnpm_lock": attrs.source(
+            doc = """Pnpm lock file""",
+        ),
+        "_python_toolchain": attrs.toolchain_dep(
+            default = "toolchains//:python",
+            providers = [PythonToolchainInfo],
+        ),
+        "_simple_pnpm_toolchain": attrs.toolchain_dep(
+            default = "toolchains//:simple_pnpm",
+            providers = [SimplePnpmToolchainInfo],
+        ),
+    },
+)
+
+def npm_bin_impl(ctx: AnalysisContext) -> list[[DefaultInfo, RunInfo, TemplatePlaceholderInfo]]:
+    bin_name = ctx.attrs.bin_name or ctx.attrs.name
+
+    exe = ctx.actions.declare_output(bin_name)
+
+    simple_pnpm_toolchain = ctx.attrs._simple_pnpm_toolchain[SimplePnpmToolchainInfo]
+
+    cmd = cmd_args(
+        ctx.attrs._python_toolchain[PythonToolchainInfo].interpreter,
+        simple_pnpm_toolchain.build_npm_bin[DefaultInfo].default_outputs,
+        "--bin-out-path",
+        exe.as_output(),
+    )
+
+    cmd.add([
+        ctx.attrs.node_modules,
+        bin_name,
+    ])
+
+    ctx.actions.run(cmd, category = "build_npm_bin", identifier = bin_name)
+
+    return [
+        DefaultInfo(default_output = exe),
+        RunInfo(exe),
+        TemplatePlaceholderInfo(
+            keyed_variables = {
+                "exe": exe,
+            },
+        ),
+    ]
+
+_npm_bin = rule(
+    impl = npm_bin_impl,
+    attrs = {
+        "bin_name": attrs.option(
+            attrs.string(),
+            default = None,
+            doc = """Node module bin name (default: attrs.name).""",
+        ),
+        "node_modules": attrs.source(
+            doc = """Target which builds `node_modules`.""",
+        ),
+        "_python_toolchain": attrs.toolchain_dep(
+            default = "toolchains//:python",
+            providers = [PythonToolchainInfo],
+        ),
+        "_simple_pnpm_toolchain": attrs.toolchain_dep(
+            default = "toolchains//:simple_pnpm",
+            providers = [SimplePnpmToolchainInfo],
+        ),
+    },
+)
+
+def npm_bin(
+        visibility = ["PUBLIC"],
+        node_modules = ":node_modules",
```

Comment:
@vindard 

I have found a legit reason to have a macro wrapper. In the `attrss ={}` section above you can only specify a `default` target as a dependency that has an aboslute path (like `toolchain//:simple_pnpm`) not a relative one (like `:node_modules` which is in the same CELL).

Still no need to have completely other file just for this IMO.